core/src/main/java/org/jruby/RubyHash.java

```diff
@@ -1581,28 +1575,59 @@ public IRubyObject each_key(final ThreadContext context, final Block block) {
         return block.isGiven() ? each_keyCommon(context, block) : enumeratorizeWithSize(context, this, "each_key", RubyHash::size);
     }
 
-    @JRubyMethod(name = "transform_keys")
-    public IRubyObject transform_keys(final ThreadContext context, final Block block) {
-        if (block.isGiven()) {
-            RubyHash result = newHash(context.runtime);
-            visitAll(context, new TransformKeysVisitor(block), result);
-            return result;
+    @JRubyMethod(name = "transform_keys", optional =  1, rest = true)
+    public IRubyObject transform_keys(final ThreadContext context, IRubyObject[] arg0, final Block block) {
+        RubyArray transformKeys = null;
+        RubyHash result = newHash(context.runtime);
+
+        if((arg0.length > 0) && (arg0 [0] != null))
+            transformKeys = ((RubyHash) arg0[0]).keys();
+
+        if(block.isGiven())
+            visitAll(context, new TransformKeysVisitor(block, transformKeys), result);
+        else
+            result = this;
+
+        if(transformKeys != null) {
+            for (int i = 0; i < transformKeys.size(); i++) {
+                RubySymbol orignalKey = (RubySymbol) transformKeys.get(i);
+                IRubyObject newKeySym = (IRubyObject) ((RubyHash) arg0[0]).get(orignalKey);
```

Comment:
I reread this a couple of times.  It feels like arg0[0] here should have a meaningful name like you did with arg0[0].keys().